tracer/src/Datadog.Trace/AppSec/WafEncoding/Obj.cs

```diff
@@ -72,32 +66,8 @@ public DdwafObjectStruct InnerStruct
 
         public IntPtr RawPtr => ptr;
 
-        public void Dispose(WafLibraryInvoker libraryInvoker)
-        {
-            if (libraryInvoker != null)
-            {
-                var rawPtr = ptr;
-                libraryInvoker.ObjectFreePtr(ref rawPtr);
-                Dispose();
-            }
-        }
-
         public void Dispose()
         {
-            Dispose(true);
-            GC.SuppressFinalize(this);
-        }
-
-        private void Dispose(bool disposing)
-        {
-            if (disposed)
-            {
-                ptr = IntPtr.Zero;
-                return;
-            }
-
-            disposed = true;
-
```

Comment:
This is intended to make it thread safe right? IIUC, you now have a potential double free, where you could call `Marshal.FreeHGlobal(ptr);` twice? I think if you keep the `disposed` field you can avoid that?

Reply:
fixed in [fe857a0](https://github.com/DataDog/dd-trace-dotnet/pull/5305/commits/fe857a062dbe7d61b461f779b5c885222479748e)

Reply:
I mean the dispose should be called just once by one thread but in case..